RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h

```diff
@@ -0,0 +1,303 @@
+      // when two valid pixels within +/- 1 in x or y are found, set their id to the minimum;
+      // after the loop, all the pixel in each cluster should have the id equeal to the lowest
+      // pixel in the cluster ( clus[i] == i ).
+      bool more = true;
+      int nloops = 0;
+      while (__syncthreads_or(more)) {
+        if (1 == nloops % 2) {
+          for (auto j = threadIdx.x, k = 0U; j < hist.size(); j += blockDim.x, ++k) {
+            auto p = hist.begin() + j;
+            auto i = *p + firstPixel;
+            auto m = clusterId[i];
+            while (m != clusterId[m])
+              m = clusterId[m];
+            clusterId[i] = m;
+          }
+        } else {
+          more = false;
+          for (auto j = threadIdx.x, k = 0U; j < hist.size(); j += blockDim.x, ++k) {
+            auto p = hist.begin() + j;
+            auto i = *p + firstPixel;
+            for (int kk = 0; kk < nnn[k]; ++kk) {
+              auto l = nn[k][kk];
+              auto m = l + firstPixel;
+              assert(m != i);
+              auto old = atomicMin(&clusterId[m], clusterId[i]);
+              if (old != clusterId[i]) {
+                // end the loop only if no changes were applied
+                more = true;
+              }
+              atomicMin(&clusterId[i], old);
+            }  // nnloop
+          }    // pixel loop
+        }
+        ++nloops;
+      }  // end while
+
+#ifdef GPU_DEBUG
+      {
+        __shared__ int n0;
+        if (threadIdx.x == 0)
+          n0 = nloops;
+        __syncthreads();
+        auto ok = n0 == nloops;
+        assert(__syncthreads_and(ok));
+        if (thisModuleId % 100 == 1)
+          if (threadIdx.x == 0)
+            printf("# loops %d\n", nloops);
+      }
+#endif
+
+      __shared__ unsigned int foundClusters;
+      foundClusters = 0;
+      __syncthreads();
+
+      // find the number of different clusters, identified by a pixels with clus[i] == i;
+      // mark these pixels with a negative id.
+      for (int i = first; i < msize; i += blockDim.x) {
+        if (id[i] == invalidModuleId)  // skip invalid pixels
+          continue;
+        if (clusterId[i] == i) {
+          auto old = atomicInc(&foundClusters, 0xffffffff);
+          clusterId[i] = -(old + 1);
+        }
+      }
+      __syncthreads();
+
+      // propagate the negative id to all the pixels in the cluster.
+      for (int i = first; i < msize; i += blockDim.x) {
+        if (id[i] == invalidModuleId)  // skip invalid pixels
+          continue;
+        if (clusterId[i] >= 0) {
+          // mark each pixel in a cluster with the same id as the first one
+          clusterId[i] = clusterId[clusterId[i]];
+        }
+      }
+      __syncthreads();
+
+      // adjust the cluster id to be a positive value starting from 0
+      for (int i = first; i < msize; i += blockDim.x) {
+        if (id[i] == invalidModuleId) {  // skip invalid pixels
+          clusterId[i] = -9999;
```

Comment:
I've changed this back to `-9999`.
@VinInn @makortel if you have suggestion for a better value and/or named constant, let me know.